Check cloud feature before setting billing access for web

kimlisa opened the pull request:
fixes https://github.com/gravitational/teleport/issues/6447

#### Description
Init web handler with feature flags on proxy init
- Retrieve auth server settings by calling `Ping`, which contains the server feature flags in the response
- Check for cloud feature is supported before setting billing access

kimlisa committed:
Check cloud feature before setting billing access for web

alex-kovoy commented on lib/web/apiserver.go:
```diff
@@ -498,7 +498,12 @@ func (h *Handler) getUserContext(w http.ResponseWriter, r *http.Request, p httpr
 	}
 
-	userContext, err := ui.NewUserContext(user, roleset)
+	pingResponse, err := clt.Ping(r.Context())
```
Is this new API method? Did it exist in 6.0 ?

kimlisa replied:
The method existed since last year, but sending back the `features` was added in `6.1`

alex-kovoy replied:
@kimlisa can we call this method once upon `NewHandler` initialization? This way we do not need to do make an extra call each time Web UI loads.

Also, since it was added recently, will it break if ping method does not have this struct?

alex-kovoy reviewed (commented):
is this API

kimlisa committed:
Retrieve auth server features on register.proxy
- Use Ping (that responds with server features) instead of GetNamespace to test c/n
- Set server features for web handler (to set it once on register.proxy)

alex-kovoy commented on lib/web/apiserver.go:
```diff
@@ -498,7 +506,9 @@ func (h *Handler) getUserContext(w http.ResponseWriter, r *http.Request, p httpr
 	}
 
-	userContext, err := ui.NewUserContext(user, roleset)
+	fmt.Println("--------------------- >>>> : ", h.authServerFeatures)
```
> fmt.Println("--------------------- >>>> : ", h.authServerFeatur

some leftovers

alex-kovoy reviewed (approved):
Looks good to me, just do not forget to remove Print statements.

kimlisa committed:
Address CRs
- Add lock before setting
- Remove print statements

r0mant commented on lib/service/connect.go:
```diff
@@ -55,8 +56,9 @@ func (process *TeleportProcess) reconnectToAuthService(role teleport.Role) (*Con
 			// client works, by using call that should succeed at all times
 			if connector.Client != nil {
-				_, err = connector.Client.GetNamespace(defaults.Namespace)
+				pingResponse, err := connector.Client.Ping(context.TODO())
```
Use `process.ExitContext()` here instead of `context.TODO()`.

kimlisa replied:
revised

r0mant commented on lib/service/connect.go:
```diff
@@ -55,8 +56,9 @@ func (process *TeleportProcess) reconnectToAuthService(role teleport.Role) (*Con
-				_, err = connector.Client.GetNamespace(defaults.Namespace)
+				pingResponse, err := connector.Client.Ping(context.TODO())
 				if err == nil {
+					process.setAuthServerFeatures(*pingResponse.GetServerFeatures())
```
`pingResponse.GetServerFeatures()` can return nil and then this will panic.

kimlisa replied:
checked for nil before dereferencing

r0mant commented on lib/service/cfg.go:
```diff
@@ -204,6 +205,9 @@ type Config struct {
 	PluginRegistry plugin.Registry
+
+	// AuthServerFeatures contain flags for supported and unsupported features.
+	AuthServerFeatures proto.Features
```
Nit: I think AuthServerFeatures is a bit too specific. How about ClusterFeatures? If you decide to rename, let's rename all authServerFeatures below as well.

kimlisa replied:
renamed

r0mant commented on lib/service/service.go:
```diff
@@ -355,6 +356,13 @@ func (process *TeleportProcess) addConnector(connector *Connector) {
+	process.Lock()
+	defer process.Unlock()
+
+	process.Config.AuthServerFeatures = features
```
Hmm, I'm not sure why this sets it in the `process.Config`? I would probably make an internal field TeleportProcess instead and initialize that (similar to `process.connectors` above).

kimlisa replied:
revised

r0mant commented on lib/service/service.go:
```diff
@@ -355,6 +356,13 @@ func (process *TeleportProcess) addConnector(connector *Connector) {
 	process.connectors[connector.ClientIdentity.ID.Role] = connector
 }
 
+func (process *TeleportProcess) setAuthServerFeatures(features proto.Features) {
```
You also need a counterpart getter that returns features under process.Lock() and use that instead of accessing them directly.

kimlisa replied:
defined a getter

kimlisa committed:
Address CR
- rename authServerFeatures to clusterFeatures
- defined a locked getter to retrieve cluster features
- check for nil before setting

r0mant commented on lib/service/connect.go:
```diff
@@ -55,8 +55,9 @@ func (process *TeleportProcess) reconnectToAuthService(role teleport.Role) (*Con
-				_, err = connector.Client.GetNamespace(defaults.Namespace)
+				pingResponse, err := connector.Client.Ping(process.ExitContext())
 				if err == nil {
+					process.setClusterFeatures(pingResponse.GetServerFeatures())
```
Could you please log server features loaded from auth server here? In Info mode. It would be helpful for troubleshooting, right now there's no visibility.

kimlisa replied:
logged

russjones reviewed (approved):
Bot.

kimlisa committed:
Add info log on features received from auth server

kimlisa committed:
Merge branch 'master' into lisa/cloud-feature-flag

kimlisa committed:
no diff
